### PR TITLE
Added a v2 parameter to dropout APIs; drop in training, scale in eval

### LINKED ISSUE #477: Dropout mode
I'm trying to import some models from theano to pytorch, but I'm facing issues with the dropout, which has a different mode: instead of doing nothing in eval and dropping + scaling in training, in theano I had scaling in eval and dropping in training.

Torch had this option to create the dropout module in one mode or the other. Could we also have this in pytorch?

### PR DESCRIPTION
**Pull request by goelhardik**
Fixes #477.

### CODE REVIEW

**Comment by fmassa** [comment]
Note that you didn't add documentation for the new argument.
Also, I think that this is a legacy mode for dropout, and it might be better handled via the approach described in #477 ?

**Comment by colesbury** [comment]
Importing legacy behavior from Torch doesn't seem like a good idea to me

**Comment by soumith** [comment]
v2 is the wrong word, but this param is helpful to have and gives the needed behavior for users.
The param has to be changed to: scale_train=True, and scale_train=False enables the v1 behavior.

**Comment by goelhardik** [comment]
So do you guys mean keep a `scale_train=True` default parameter and use `y = F.dropout(x,p,mode) * (1-p)` in `nn.Dropout` to fix this?

**Comment by soumith** [comment]
Right now, we scale down `y` by `(1-p)` in training mode via this piece of code: `.div_(1 - self.p)`.

This is done so that models at test time do not need to have scaling (and can entirely remove the dropout layer).

`v1` behavior is to not scale anything in training mode but instead scale in test mode.

Keep current behavior as-is with `scale_train=True` (default value). and if `scale_train=False`, then instead do no scaling in train mode, but do scaling in test mode.

File: torch/nn/modules/dropout.py [hunk 1]
```diff
@@ -21,22 +24,25 @@ class Dropout(Module):
...
+    def __init__(self, p=0.5, inplace=False, scale_train=True):
         super(Dropout, self).__init__()
         if p < 0 or p > 1:
             raise ValueError("dropout probability has to be between 0 and 1, "
                              "but got {}".format(p))
         self.p = p
         self.inplace = inplace
+        self.scale_train = scale_train
 
     def forward(self, input):
```
Comment (apaszke) [inline torch/nn/modules/dropout.py]: You need to add `__setstate__` so that it initializes `self.scale_train` with a default value. Otherwise all checkpoints containing dropout modules will be broken.

Comment (goelhardik) [inline torch/nn/modules/dropout.py]: Do you mean have something like this?
`def __setstate__(self, state):
     if 'scale_train' not in state:
         self.__dict__.update({'scale_train' : True})`

Comment (apaszke) [inline torch/nn/modules/dropout.py]: Or better:
```python
def __setstate__(self, state):
    state.setdefault('scale_train', True)
    self.__dict__ = state
```

**Comment by goelhardik** [comment]
Looks like `test_Softshrink_lambda` is failing only for Python 3.6 in the Travis build. Any idea why?

File: torch/nn/_functions/dropout.py [hunk 2]
```diff
@@ -24,21 +25,28 @@ def forward(self, input):
...
-                self.noise.fill_(0)
-            self.noise = self.noise.expand_as(input)
-            output.mul_(self.noise)
+        if self.p > 0:
+            if self.train:
+                self.noise = self._make_noise(input)
+                self.noise.bernoulli_(1 - self.p)
+                if self.scale_train:
+                    self.noise.div_(1 - self.p)
+                if self.p == 1:
```
Comment (apaszke) [inline torch/nn/_functions/dropout.py]: can you move this up and make it (right now we unnecessarily call `bernoulli_` when `p == 1`)
```python
if self.p == 1:
    self.noise.fill_(0)
else:
    self.noise.bernoulli_(1 - self.p)
    if self.scale_train:
        self.noise.div_(1 - self.p)
```

**Review by apaszke — APPROVED** [review]
This looks good, can you only add that single improvement? Thanks!

**Comment by soumith** [comment]
this PR has rotted for too long, and the feature i think is not really worth adding, as we dont have compelling torch transitions anymore.

I'm really sorry for letting this rot @goelhardik and thanks a lot for the PR!
